### PR TITLE
 role base access controll implemented

### PR DESCRIPTION
**Pull request by Deepak-webelight**
- added role feild in usermodel 
- implemented auth guard for authentication
- similarly implemented role guard to access rbac

### CODE REVIEW

File: server/src/modules/user/user.dto.ts [hunk 1]
```diff
@@ -16,6 +18,11 @@ export class SignUpRequestBodyDto {
   @IsNotEmpty()
   @IsStrongPassword()
   readonly password: string;
+
+  @IsNotEmpty()
+  @IsEnum(Role)
+  @IsString()
+  readonly role: keyof typeof Role;
```
Comment (anshs-webelight) [inline server/src/modules/user/user.dto.ts]: why do we accept a role from frontend while signing up?, we can't create an admin user, only the admin can create that

Comment (anshs-webelight) [reply server/src/modules/user/user.dto.ts]: while signing up role should be user-only, and need to create default entry for admin user, we can't create through sign up

File: server/src/modules/user/user.model.ts [hunk 2]
```diff
@@ -17,6 +18,12 @@ export class User {
     required: true,
   })
   password: string;
+
+  @Prop({
+    required: true,
+    enum: Role,
```
Comment (anshs-webelight) [inline server/src/modules/user/user.model.ts:24]: give default value as well

File: server/src/modules/user/user.roleGuard.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common';
+import { SetMetadata } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { ROLES_KEY } from 'src/utils/constants';
+
+export enum Role {
+  User = 'User',
+  Admin = 'Admin',
+}
```
Comment (anshs-webelight) [inline server/src/modules/user/user.roleGuard.ts]:
```suggestion
export enum Role {
  User,
  Admin,
}
```

Comment (anshs-webelight) [reply server/src/modules/user/user.roleGuard.ts]: Means,
```
role: 0 ==>   User 
role: 1 ==>  Admin 
```

File: server/src/modules/user/user.roleGuard.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+// Role decorator
+export const Roles = (...roles: Role[]) => SetMetadata(ROLES_KEY, roles);
```
Comment (anshs-webelight) [inline server/src/modules/user/user.roleGuard.ts]: why do we add a roles array?

File: server/src/modules/user/user.roleGuard.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        ROLES_KEY,
+        [context.getHandler(), context.getClass()],
+      );
+      if (!requiredRoles) {
+        return true;
+      }
+      const { user } = context.switchToHttp().getRequest();
+      return requiredRoles[0] === user.role;
+    } catch (err) {
+      throw new Error((err as Error).message);
```
Comment (anshs-webelight) [inline server/src/modules/user/user.roleGuard.ts]: throw unAuthorized error

File: server/src/modules/user/user.authGuard.ts
Comment (anshs-webelight) [inline server/src/modules/user/user.authGuard.ts:1]: folder should be `src/guards/auth.guard.ts`

File: server/src/appConfig/configuration.validate.ts [hunk 6]
```diff
@@ -6,4 +6,5 @@ export default Joi.object({
   JWT_SECRET: Joi.string().required(),
   TOKEN_EXPIRY: Joi.alternatives(Joi.string(), Joi.number()).required(),
   REFRESH_TOKEN_EXPIRY: Joi.alternatives(Joi.string(), Joi.number()).required(),
+
```
Comment (anshs-webelight) [inline server/src/appConfig/configuration.validate.ts]: remove extra line

File: server/src/appConfig/configuration.validate.ts [hunk 7]
```diff
@@ -6,4 +6,5 @@ export default Joi.object({
   JWT_SECRET: Joi.string().required(),
```
Comment (anshs-webelight) [inline server/src/appConfig/configuration.validate.ts]:
```suggestion
  import { object, string, number, alternatives } from 'joi';

export default object({
  MONGODB_URL: string().required(),
  PORT: number().required().default(5000),
  JWT_SECRET: string().required(),
  TOKEN_EXPIRY: alternatives(string(), number()).required(),
  REFRESH_TOKEN_EXPIRY: alternatives(string(), number()).required(),
});

```

Comment (anshs-webelight) [reply server/src/appConfig/configuration.validate.ts]: don't use `* as Joi` syntax for imports

File: server/src/modules/user/user.roleGuard.ts
Comment (anshs-webelight) [inline server/src/modules/user/user.roleGuard.ts:1]: file should be at `src/guards/role-auth.guard.ts`

File: server/src/modules/user/user.authGuard.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      return true;
+    } catch (error) {
+      throw new UnauthorizedException('Expired token | Not found');
+    }
+  }
+
+  private extractTokenFromHeader(request: Request): string | undefined {
+    const [type, token] =
+      (request.cookies.token || request.cookies.refreshToken)?.split(' ') ?? [];
+    return type === 'Bearer' ? token || false : undefined;
```
Comment (anshs-webelight) [inline server/src/modules/user/user.authGuard.ts]: why do we need to return a falsy value if not token? can't we just throw error here?

File: server/src/modules/user/user.authGuard.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    // switch context to request
+    const req: Request = context.switchToHttp().getRequest();
+
+    try {
+      //   Extract token from request
+      const token = this.extractTokenFromHeader(req);
+
+      if (!token) {
+        throw new UnauthorizedException();
```
Comment (anshs-webelight) [inline server/src/modules/user/user.authGuard.ts]: add proper message

File: server/src/modules/user/user.authGuard.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      }
+
+      //  veryfy token
+      const verifyToken = await this.jwtService.verify(token);
+
+      req['user'] = verifyToken;
+      return true;
+    } catch (error) {
+      throw new UnauthorizedException('Expired token | Not found');
```
Comment (anshs-webelight) [inline server/src/modules/user/user.authGuard.ts]:
```suggestion
      throw new UnauthorizedException('Access Denied: invalid token');
```

File: server/src/modules/user/user.dto.ts [hunk 11]
```diff
@@ -16,6 +18,11 @@ export class SignUpRequestBodyDto {
   @IsNotEmpty()
   @IsStrongPassword()
   readonly password: string;
+
+  @IsNotEmpty()
+  @IsEnum(Role)
+  @IsString()
```
Comment (anshs-webelight) [inline server/src/modules/user/user.dto.ts]: @IsEnum(Role) and  @IsString() not needed

File: server/src/modules/user/user.roleGuard.ts (hunk 3 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.roleGuard.ts]: okay understood, and implemented

File: server/src/modules/user/user.dto.ts (hunk 1 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.dto.ts]: Implemented, 
- By default user will be sign up for user role only.
- for admin created a new route /admin which requires request body same as /sign-up, additionally required x-api-key in header

File: server/src/modules/user/user.model.ts (hunk 2 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.model.ts:24]: given

File: server/src/modules/user/user.roleGuard.ts (hunk 4 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.roleGuard.ts]: updated

File: server/src/appConfig/configuration.validate.ts (hunk 7 above)
Comment (Deepak-webelight) [reply server/src/appConfig/configuration.validate.ts]: updated

File: server/src/modules/user/user.authGuard.ts (hunk 8 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.authGuard.ts]: updated

File: server/src/modules/user/user.authGuard.ts (hunk 9 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.authGuard.ts]: updated

File: server/src/guards/role-auth.guard.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+import { Reflector } from '@nestjs/core';
+import appConfig from 'src/appConfig/configuration';
+
+export enum Role {
+  User,
+  Admin,
+}
+
+// Role decorator
+export const Roles = (roles: Role) => SetMetadata(appConfig('ROLES_KEY'), roles);
```
Comment (anshs-webelight) [inline server/src/guards/role-auth.guard.ts]:
```suggestion
export const Roles = (role: Role) => SetMetadata(appConfig('ROLES_KEY'), role);
```

File: server/src/guards/role-auth.guard.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+@Injectable()
+export class RolesGuard implements CanActivate {
+  constructor(private readonly reflector: Reflector) {}
+
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    try {
+      const requiredRoles = this.reflector.getAllAndOverride<Role[]>(
+        appConfig('ROLES_KEY'),
```
Comment (anshs-webelight) [inline server/src/guards/role-auth.guard.ts]: wrong implementation for appConfig file

File: server/src/modules/user/user.controller.ts [hunk 14]
```diff
@@ -112,4 +124,46 @@ export class UserController {
...
+    @Headers() headers: HeaderDto,
+    @Body() body: SignUpRequestBodyDto,
+    @Res({ passthrough: true }) res: Response,
+  ): Promise<IUserResponse> {
+    try {
+      const xApiKey = headers['x-api-key'];
+
+      if (!xApiKey)
+        throw new UnauthorizedException('x-api-key key not provided');
+
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: we should have super admin auth guard for this

File: server/src/modules/user/user.controller.ts [hunk 15]
```diff
@@ -112,4 +124,46 @@ export class UserController {
...
+      // call User Service to varify and resister admin user
+      const { _id, role } = await this.userService.registerAdminUser(
+        xApiKey,
+        body,
+      );
+
+      const { token, refreshToken } = await this.userService.generateTokens(
+        _id,
+        role,
+      );
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: call the single service file's method and do all the processes into service file

File: server/src/modules/user/user.controller.ts [hunk 16]
```diff
@@ -112,4 +124,46 @@ export class UserController {
...
+
+      res.cookie('token', tokenFormat(token), {
+        httpOnly: true,
+        sameSite: 'strict',
+      });
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: create const variable for `{ httpOnly: true, samSite: "strict" }` and use for both

File: server/src/modules/user/user.interface.ts [hunk 17]
```diff
@@ -1,9 +1,13 @@
...
-  port: number;
-  database: string;
-  jwtSecret: string;
-  tokenExpiry: string | number;
-  refreshTokenExpiry: string | number;
+  PORT: number;
+  MONGODB_URL: string;
+  JWT_SECRET: string;
+  TOKEN_EXPIRY: string | number;
+  REFRESH_TOKEN_EXPIRY: string | number;
```
Comment (anshs-webelight) [inline server/src/modules/user/user.interface.ts]: this both should be come from constant file not from env variables

File: server/src/modules/user/user.service.ts [hunk 18]
```diff
@@ -33,6 +32,7 @@ export class UserService {
         email,
         name,
         password: userHashedPassword,
+        role: Role.User, // by default user role only
```
Comment (anshs-webelight) [inline server/src/modules/user/user.service.ts]: no need to pass role as we have defined default value for role

File: server/src/guards/role-auth.guard.ts (hunk 13 above)
Comment (Deepak-webelight) [reply server/src/guards/role-auth.guard.ts]: fixed using this pattern 

```typescript
 
import { config } from 'dotenv';
config();
export const appConfig = {
  port: process.env.PORT,
  mongodbUrl: process.env.MONGODB_URL,
  jwtSecret: process.env.JWT_SECRET,
  xApiKey: process.env.X_API_KEY,
};
```
and how it is simply accessible by `appConfig.['variable name']`

File: server/src/modules/user/user.controller.ts (hunk 14 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.controller.ts]: implemented new SuperAdminAuthGuard for this to validate admin routes

File: server/src/modules/user/user.controller.ts (hunk 15 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.controller.ts]: updated also for signup and login

File: server/src/modules/user/user.controller.ts (hunk 16 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.controller.ts]: updated

File: server/src/modules/user/user.interface.ts (hunk 17 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.interface.ts]: updated

File: server/src/modules/user/user.controller.ts [hunk 19]
```diff
@@ -112,4 +105,32 @@ export class UserController {
       throw new BadRequestException(error.response);
     }
   }
+
+  @PublicRoute()
+  @SuperAdmin()
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts:110]: why do we need these two decorators? we can directly use `@UseGuards(SuperAdminAuthGuard)`

File: server/src/modules/user/user.interface.ts [hunk 20]
```diff
@@ -1,10 +1,3 @@
-export interface Iconfiguration {
-  port: number;
-  database: string;
-  jwtSecret: string;
-  tokenExpiry: string | number;
-  refreshTokenExpiry: string | number;
-}
 export interface IUserResponse {
   message: string;
   data?: any;
```
Comment (anshs-webelight) [inline server/src/modules/user/user.interface.ts:3]: don't use any as type explore about generic types

File: server/src/guards/superAdmin.auth.guard.ts [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+  ExecutionContext,
+  Injectable,
+  SetMetadata,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { appConfig } from 'src/appConfig/configuration';
+import { SuperAdminKey } from 'src/utils/constants';
+
+export const SuperAdmin = () => SetMetadata(SuperAdminKey, true);
```
Comment (anshs-webelight) [inline server/src/guards/superAdmin.auth.guard.ts:12]: not needed

File: server/src/guards/superAdmin.auth.guard.ts [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+export class SuperAdminAuthGuard implements CanActivate {
+  constructor(private readonly reflector: Reflector) {}
+
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    try {
+      const isSuperAdmin = this.reflector.getAllAndOverride<boolean>(
+        SuperAdminKey,
+        [context.getHandler(), context.getClass()],
+      );
+      if (!isSuperAdmin) return true;
```
Comment (anshs-webelight) [inline server/src/guards/superAdmin.auth.guard.ts:24]: why are we returning true without checking api-key header?

File: server/src/guards/superAdmin.auth.guard.ts [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+      const req = context.switchToHttp().getRequest();
+      const ReqxApiKey = req.headers['x-api-key'];
+
+      if (ReqxApiKey !== appConfig.xApiKey) {
```
Comment (anshs-webelight) [inline server/src/guards/superAdmin.auth.guard.ts:29]: ReqApiKey ==> apiKey

File: server/src/guards/superAdmin.auth.guard.ts [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+        return false;
+      }
+      return true;
```
Comment (anshs-webelight) [inline server/src/guards/superAdmin.auth.guard.ts:32]: if (ReqxApiKey === appConfig.xApiKey) {
   return true
}
return false;

File: server/src/appConfig/configuration.ts [hunk 25]
```diff
@@ -1,9 +1,16 @@
...
-  port: parseInt(process.env.PORT) || 5000,
-  database: process.env.MONGODB_URL,
+export const appConfig = {
+  port: process.env.PORT,
+  mongodbUrl: process.env.MONGODB_URL,
   jwtSecret: process.env.JWT_SECRET,
-  tokenExpiry: process.env.TOKEN_EXPIRY,
-  refreshTokenExpiry: process.env.REFRESH_TOKEN_EXPIRY,
-});
+  xApiKey: process.env.X_API_KEY,
```
Comment (anshs-webelight) [inline server/src/appConfig/configuration.ts:9]: give a proper name like `adminApiKey: process.env.ADMIN_API_KEY`

File: server/src/modules/user/user.controller.ts (hunk 19 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.controller.ts:110]: implement `@UseGuards(SuperAdminAuthGuard)` as discussed earlier, and `@publicRoute()` decorator is required as AuthGuard is working globally thus its token checking should not work for `/admin` route implemented

File: server/src/modules/user/user.interface.ts (hunk 20 above)
Comment (Deepak-webelight) [reply server/src/modules/user/user.interface.ts:3]: implemented

File: server/src/guards/superAdmin.auth.guard.ts (hunk 22 above)
Comment (Deepak-webelight) [reply server/src/guards/superAdmin.auth.guard.ts:24]: logic corrected

**Pull request merged**